# Before you start :
   - Keep in mind that you need to use some of the functions you learned in the previous lessons.
   - All datasets are provided in the IronHack database.
   - Elaborate your codes and outputs as much as you can.
   - Try your best to answer the questions and complete the tasks and most importantly enjoy the process!!!

#### Import all the libraries that are necessary

In [2]:
import numpy as np
import pandas as pd

# Challenge 1

#### In this challenge we will use the `Temp_States`  database. 

#### First import it into a data frame called `temp`.

In [3]:
temp = pd.read_csv('states.csv')

#### Print `temp`

In [ ]:
temp.drop('Unnamed: 0', axis=1, inplace=True)

In [14]:
temp

City        State  Temperature
0         NYC     New York    19.444444
1      Albany     New York     9.444444
2     Buffalo     New York     3.333333
3    Hartford  Connecticut    17.222222
4  Bridgeport  Connecticut    14.444444
5      Treton   New Jersey    22.222222
6      Newark   New Jersey    20.000000

#### Explore the data types of the Temp dataframe. What type of data do we have? Comment your result.

In [16]:
temp.dtypes

City            object
State           object
Temperature    float64
dtype: object

The data types are object for City and State (which is the equivalent of string in python) and float64 for Temperature. Therefore nothing needs to be changed.

#### Select the rows where state is New York

In [18]:
temp[temp['State']=='New York']

City     State  Temperature
0      NYC  New York    19.444444
1   Albany  New York     9.444444
2  Buffalo  New York     3.333333

#### What is the average of the temperature of cities in New York?

In [19]:
temp[temp['State']=='New York'].mean()

Temperature    10.740741
dtype: float64

#### We want to know cities and states with Temperature above 15 degress Celcius

In [21]:
temp[temp['Temperature']>15]

City        State  Temperature
0       NYC     New York    19.444444
3  Hartford  Connecticut    17.222222
5    Treton   New Jersey    22.222222
6    Newark   New Jersey    20.000000

#### Now, return only the cities that have a temperature above 15 degress Celcius

In [24]:
temp[temp['Temperature']>15][['City']]

City
0       NYC
3  Hartford
5    Treton
6    Newark

#### We want to know which cities have a temperature above 15 degrees Celcius and below 20 degrees Celcius

*Hint: First write the condition then select the rows.*

In [30]:
temp[(temp['Temperature']>15) & (temp['Temperature']<20)]

City        State  Temperature
0       NYC     New York    19.444444
3  Hartford  Connecticut    17.222222

#### Find the mean and the standard deviation of the temperature of each state.

*Hint: Use functions from Data Manipulation lesson*

In [32]:
temp.groupby('State').mean()

Temperature
State                   
Connecticut    15.833333
New Jersey     21.111111
New York       10.740741

In [33]:
temp.groupby('State').std()

Temperature
State                   
Connecticut     1.964186
New Jersey      1.571348
New York        8.133404

# Challenge 2

#### Load the `employees` database into a DataFrame. Call the dataframe `employees`

In [34]:
employees = pd.read_csv('employees.csv')

In [37]:
employees.drop('Unnamed: 0', axis=1, inplace=True)

In [38]:
employees.head()

Name Department Education Gender      Title  Years  Salary
0    Jose         IT  Bachelor      M    analyst      1      35
1   Maria         IT    Master      F    analyst      2      30
2   David         HR    Master      M    analyst      2      30
3   Sonia         HR  Bachelor      F    analyst      4      35
4  Samuel      Sales    Master      M  associate      3      55

#### Explore the data types of the Temp dataframe. Comment your results

In [39]:
employees.dtypes

Name          object
Department    object
Education     object
Gender        object
Title         object
Years          int64
Salary         int64
dtype: object

The data types are object for all text fields except for the numeric fields (years and salary), which are int64.

#### What's the average salary in this company?

In [40]:
employees['Salary'].mean()

48.888888888888886

#### What's the highest salary?

In [41]:
employees['Salary'].max()

70

#### What's the lowest salary?

In [42]:
employees['Salary'].min()

30

#### Who are the employees with the lowest salary?

In [43]:
employees[employees['Salary'] == employees['Salary'].min()]

Name Department Education Gender    Title  Years  Salary
1  Maria         IT    Master      F  analyst      2      30
2  David         HR    Master      M  analyst      2      30

#### Could you give all the information about an employee called David?

In [44]:
employees[employees['Name']=='David']

Name Department Education Gender    Title  Years  Salary
2  David         HR    Master      M  analyst      2      30

#### Could you give only David's salary?

In [46]:
employees[employees['Name']=='David'][['Salary']]

Salary
2      30

#### Print all the rows where job title is associate

In [47]:
employees[employees['Title']=='associate']

Name Department Education Gender      Title  Years  Salary
4  Samuel      Sales    Master      M  associate      3      55
5     Eva      Sales  Bachelor      F  associate      2      55
7   Pedro         IT       Phd      M  associate      7      60

#### Print the first 3 rows of your dataframe

##### Tip : There are 2 ways to do it. Do it both ways

In [48]:
employees.head(3)

Name Department Education Gender    Title  Years  Salary
0   Jose         IT  Bachelor      M  analyst      1      35
1  Maria         IT    Master      F  analyst      2      30
2  David         HR    Master      M  analyst      2      30

In [52]:
employees.iloc[:3]

Name Department Education Gender    Title  Years  Salary
0   Jose         IT  Bachelor      M  analyst      1      35
1  Maria         IT    Master      F  analyst      2      30
2  David         HR    Master      M  analyst      2      30

#### Find the employees who's title is associate and the salary above 55?

In [53]:
employees[(employees['Title']=='associate') & (employees['Salary'] > 55)]

Name Department Education Gender      Title  Years  Salary
7  Pedro         IT       Phd      M  associate      7      60

#### Group the employees based on their number of years of employment. What are the average salaries in each group?

In [55]:
employees.groupby('Years').mean()

Salary
Years           
1      35.000000
2      38.333333
3      55.000000
4      35.000000
7      60.000000
8      70.000000

####  What is the average Salary per title?

In [58]:
employees.groupby('Title').mean()[['Salary']]

Salary
Title               
VP         70.000000
analyst    32.500000
associate  56.666667

####  What are the Salary quartiles?


In [59]:
employees['Salary'].quantile(q=0.25)

35.0

In [60]:
employees['Salary'].quantile(q=0.75)

60.0

#### Is the mean salary per gender different?

In [62]:
employees[employees['Gender']=='M'].mean()['Salary']

50.0

In [63]:
employees[employees['Gender']=='F'].mean()['Salary']

47.5

#### Find the minimum, mean and the maximum of all numeric columns for each Department.



In [65]:
employees[['Years', 'Salary']].min()

Years      1
Salary    30
dtype: int64

In [68]:
employees[['Years', 'Salary']].max()

Years      8
Salary    70
dtype: int64

In [69]:
employees[['Years', 'Salary']].mean()

Years      4.111111
Salary    48.888889
dtype: float64

### Bonus Question

#### For each department, compute the difference between the maximal salary and the minimal salary.

##### * Hint: try using `agg` or `apply` and `lambda`*

In [70]:
employees.head()

Name Department Education Gender      Title  Years  Salary
0    Jose         IT  Bachelor      M    analyst      1      35
1   Maria         IT    Master      F    analyst      2      30
2   David         HR    Master      M    analyst      2      30
3   Sonia         HR  Bachelor      F    analyst      4      35
4  Samuel      Sales    Master      M  associate      3      55

In [75]:
employees_diff = employees.groupby('Department').agg({'Salary' : ['max', 'min']})
employees_diff

Salary    
              max min
Department           
HR             70  30
IT             70  30
Sales          55  55

In [77]:
employees_diff['Salary']['max']

Department
HR       70
IT       70
Sales    55
Name: max, dtype: int64

In [79]:
employees_diff['diff'] = employees_diff['Salary']['max'] - employees_diff['Salary']['min']

In [80]:
employees_diff

Salary     diff
              max min     
Department                
HR             70  30   40
IT             70  30   40
Sales          55  55    0

# Challenge 3

#### Open the Orders dataset. Name your dataset orders

In [81]:
orders = pd.read_csv('orders.csv')

In [83]:
orders.drop('Unnamed: 0', axis=1, inplace=True)

In [84]:
orders.head()

index  InvoiceNo StockCode  year  month  day  hour  \
0      0     536365    85123A  2010     12    3     8   
1      1     536365     71053  2010     12    3     8   
2      2     536365    84406B  2010     12    3     8   
3      3     536365    84029G  2010     12    3     8   
4      4     536365    84029E  2010     12    3     8   

                           Description  Quantity          InvoiceDate  \
0   white hanging heart t-light holder         6  2010-12-01 08:26:00   
1                  white metal lantern         6  2010-12-01 08:26:00   
2       cream cupid hearts coat hanger         8  2010-12-01 08:26:00   
3  knitted union flag hot water bottle         6  2010-12-01 08:26:00   
4       red woolly hottie white heart.         6  2010-12-01 08:26:00   

   UnitPrice  CustomerID         Country  amount_spent  
0       2.55       17850  United Kingdom         15.30  
1       3.39       17850  United Kingdom         20.34  
2       2.75       17850  United Kingdom         22.00  
3       3.39       17850  United Kingdom         20.34  
4       3.39       17850  United Kingdom         20.34

#### Explore your dataset by looking at the data types and the summary statistics. Comment your results

In [86]:
orders.dtypes

index             int64
InvoiceNo         int64
StockCode        object
year              int64
month             int64
day               int64
hour              int64
Description      object
Quantity          int64
InvoiceDate      object
UnitPrice       float64
CustomerID        int64
Country          object
amount_spent    float64
dtype: object

In [89]:
orders.describe()

index      InvoiceNo           year          month  \
count  397924.000000  397924.000000  397924.000000  397924.000000   
mean   198961.500000  560617.126645    2010.934259       7.612537   
std    114870.908596   13106.167695       0.247829       3.416527   
min         0.000000  536365.000000    2010.000000       1.000000   
25%     99480.750000  549234.000000    2011.000000       5.000000   
50%    198961.500000  561893.000000    2011.000000       8.000000   
75%    298442.250000  572090.000000    2011.000000      11.000000   
max    397923.000000  581587.000000    2011.000000      12.000000   

                 day           hour       Quantity      UnitPrice  \
count  397924.000000  397924.000000  397924.000000  397924.000000   
mean        3.614555      12.728247      13.021823       3.116174   
std         1.928274       2.273535     180.420210      22.096788   
min         1.000000       6.000000       1.000000       0.000000   
25%         2.000000      11.000000       2.000000       1.250000   
50%         3.000000      13.000000       6.000000       1.950000   
75%         5.000000      14.000000      12.000000       3.750000   
max         7.000000      20.000000   80995.000000    8142.750000   

          CustomerID   amount_spent  
count  397924.000000  397924.000000  
mean    15294.315171      22.394749  
std      1713.169877     309.055588  
min     12346.000000       0.000000  
25%     13969.000000       4.680000  
50%     15159.000000      11.800000  
75%     16795.000000      19.800000  
max     18287.000000  168469.600000

The majority of columns are numeric: mostly integers but also a couple of floats. The rest are text (object) type.

There is a huge amount of data (n=397924). The data has been collected from the year 2010 to 2011. The mean quantity of each order is 13 items, with a mean unit price of 3.11 and an average spent amount of 22.4.

####  What is the average Purchase Price?

In [90]:
orders['amount_spent'].mean()

22.394748504739596

#### What were the highest and lowest purchase prices? 

In [91]:
orders['amount_spent'].max()

168469.6

In [92]:
orders['amount_spent'].min()

0.0

#### Select all the customers we have in Spain

In [94]:
orders[orders['Country']=='Spain']

index  InvoiceNo StockCode  year  month  day  hour  \
4250      4250     536944     22383  2010     12    5    12   
4251      4251     536944     22384  2010     12    5    12   
4252      4252     536944     20727  2010     12    5    12   
4253      4253     536944     20725  2010     12    5    12   
4254      4254     536944     20728  2010     12    5    12   
13974    13974     538095     22504  2010     12    4    14   
13975    13975     538095     22212  2010     12    4    14   
13976    13976     538095     22953  2010     12    4    14   
13977    13977     538095     84818  2010     12    4    14   
13978    13978     538095     21114  2010     12    4    14   
13979    13979     538095     84692  2010     12    4    14   
13980    13980     538095     22796  2010     12    4    14   
13981    13981     538095     40001  2010     12    4    14   
13982    13982     538095     22646  2010     12    4    14   
13983    13983     538095     21883  2010     12    4    14   
13984    13984     538095     21879  2010     12    4    14   
13985    13985     538095     22494  2010     12    4    14   
13986    13986     538095     47566  2010     12    4    14   
13987    13987     538095    85123A  2010     12    4    14   
13988    13988     538095     21430  2010     12    4    14   
13989    13989     538095     22914  2010     12    4    14   
13990    13990     538095     22913  2010     12    4    14   
13991    13991     538095     22912  2010     12    4    14   
13992    13992     538095     22631  2010     12    4    14   
13993    13993     538095     22497  2010     12    4    14   
13994    13994     538095     22909  2010     12    4    14   
13995    13995     538095     22862  2010     12    4    14   
13996    13996     538095     22726  2010     12    4    14   
13997    13997     538095     20828  2010     12    4    14   
13998    13998     538095     20829  2010     12    4    14   
...        ...        ...       ...   ...    ...  ...   ...   
394708  394708     581193     23452  2011     12    3    17   
394709  394709     581193     22907  2011     12    3    17   
394710  394710     581193     22851  2011     12    3    17   
394711  394711     581193     23209  2011     12    3    17   
394712  394712     581193     22662  2011     12    3    17   
394713  394713     581193     23583  2011     12    3    17   
394714  394714     581193     22896  2011     12    3    17   
394715  394715     581193     23203  2011     12    3    17   
394716  394716     581193     22393  2011     12    3    17   
394717  394717     581193     22391  2011     12    3    17   
394718  394718     581193     23454  2011     12    3    17   
394719  394719     581193    84536A  2011     12    3    17   
394720  394720     581193    84535B  2011     12    3    17   
394721  394721     581193     22755  2011     12    3    17   
394722  394722     581193     22753  2011     12    3    17   
394723  394723     581193     22754  2011     12    3    17   
394724  394724     581193     23372  2011     12    3    17   
394725  394725     581193     23371  2011     12    3    17   
394726  394726     581193     23254  2011     12    3    17   
394727  394727     581193     22326  2011     12    3    17   
394728  394728     581193     22629  2011     12    3    17   
394729  394729     581193     22631  2011     12    3    17   
394730  394730     581193     23290  2011     12    3    17   
394731  394731     581193     23289  2011     12    3    17   
394732  394732     581193     23292  2011     12    3    17   
394733  394733     581193     23291  2011     12    3    17   
394734  394734     581193    85232D  2011     12    3    17   
394735  394735     581193     22721  2011     12    3    17   
394736  394736     581193     23241  2011     12    3    17   
394737  394737     581193     23247  2011     12    3    17   

                                Description  Quantity          InvoiceDate  \
4250                 lun

#### How many customers do we have in Spain?
##### Hint : Use value_counts() method

In [98]:
orders['Country'].value_counts()['Spain']

2485

#### Select all the customers who have bought more than 50 items ?

In [99]:
orders[orders['Quantity']>50]

index  InvoiceNo StockCode  year  month  day  hour  \
46          46     536371     22086  2010     12    3     9   
83          83     536376     21733  2010     12    3     9   
96          96     536378     21212  2010     12    3     9   
102        102     536378    85071B  2010     12    3     9   
174        174     536386    85099C  2010     12    3     9   
175        175     536386    85099B  2010     12    3     9   
176        176     536387     79321  2010     12    3     9   
177        177     536387     22780  2010     12    3     9   
178        178     536387     22779  2010     12    3     9   
179        179     536387     22466  2010     12    3     9   
180        180     536387     21731  2010     12    3     9   
215        215     536390    84970S  2010     12    3    10   
217        217     536390     20668  2010     12    3    10   
218        218     536390    85123A  2010     12    3    10   
219        219     536390     22197  2010     12    3    10   
222        222     536390     21080  2010     12    3    10   
223        223     536390     21094  2010     12    3    10   
225        225     536390     21786  2010     12    3    10   
231        231     536390     22969  2010     12    3    10   
232        232     536390    85099B  2010     12    3    10   
245        245     536394     22633  2010     12    3    10   
246        246     536394     22866  2010     12    3    10   
247        247     536394     22865  2010     12    3    10   
248        248     536394     22632  2010     12    3    10   
376        376     536403     22867  2010     12    3    11   
406        406     536405     20914  2010     12    3    11   
673        673     536416     21915  2010     12    3    11   
718        718     536437     21154  2010     12    3    12   
719        719     536437     22189  2010     12    3    12   
720        720     536437     17021  2010     12    3    12   
...        ...        ...       ...   ...    ...  ...   ...   
397113  397113     581457     23542  2011     12    4    18   
397114  397114     581457     23531  2011     12    4    18   
397115  397115     581457     23535  2011     12    4    18   
397116  397116     581457     23530  2011     12    4    18   
397117  397117     581457     23526  2011     12    4    18   
397118  397118     581457     23401  2011     12    4    18   
397119  397119     581458     22197  2011     12    4    18   
397120  397120     581459     22197  2011     12    4    18   
397155  397155     581467     21167  2011     12    4    19   
397330  397330     581476     23167  2011     12    5     8   
397336  397336     581476     22998  2011     12    5     8   
397344  397344     581476     22693  2011     12    5     8   
397351  397351     581476     22197  2011     12    5     8   
397371  397371     581476     16014  2011     12    5     8   
397372  397372     581476     16008  2011     12    5     8   
397380  397380     581476     72232  2011     12    5     8   
397448  397448     581482     23251  2011     12    5     9   
397451  397451     581483     23843  2011     12    5     9   
397454  397454     581485     20749  2011     12    5     9   
397471  397471     581487     21137  2011     12    5     9   
397563  397563     581501     22319  2011     12    5    10   
397572  397572     581501     22544  2011     12    5    10   
397586  397586     581501     22545  2011     12    5    10   
397622  397622     581514    17091J  2011     12    5    11   
397630  397630     581514     21705  2011     12    5    11   
397720  397720     581566     23404  2011     12    5    11   
397721  397721     581567     21417  2011     12    5    11   
397729  397729     581567     21326  2011     12    5    11   
397761  397761     581571     23167  2011     12    5    12   
397882  397882     581584     20832  2011     12    5    12   

                                Description  Quantity          InvoiceDate  \
46           paper chain

#### Select orders from Spain that are above 50 items

In [100]:
orders[(orders['Country']=='Spain')&(orders['Quantity']>50)]

index  InvoiceNo StockCode  year  month  day  hour  \
4250      4250     536944     22383  2010     12    5    12   
4251      4251     536944     22384  2010     12    5    12   
4252      4252     536944     20727  2010     12    5    12   
4253      4253     536944     20725  2010     12    5    12   
4254      4254     536944     20728  2010     12    5    12   
32454    32454     540785    84997B  2011      1    2    11   
32455    32455     540785    84997D  2011      1    2    11   
32456    32456     540785    84997C  2011      1    2    11   
32457    32457     540785    84997A  2011      1    2    11   
91155    91155     547972     22383  2011      3    1    16   
91156    91156     547972     22384  2011      3    1    16   
91159    91159     547972     20728  2011      3    1    16   
174442  174442     558871     22384  2011      7    1    13   
174443  174443     558871     20727  2011      7    1    13   
174444  174444     558871     20725  2011      7    1    13   
174445  174445     558871     20728  2011      7    1    13   
174446  174446     558871    84997B  2011      7    1    13   
174447  174447     558871    84997D  2011      7    1    13   
174448  174448     558871    84997C  2011      7    1    13   
174449  174449     558871    84997A  2011      7    1    13   
197167  197167     561669     23295  2011      7    4    17   
197168  197168     561669     23297  2011      7    4    17   
197169  197169     561669     23294  2011      7    4    17   
197170  197170     561669     23296  2011      7    4    17   
197171  197171     561669     23293  2011      7    4    17   
224881  224881     564734     84826  2011      8    7    13   
254097  254097     567968     85204  2011      9    5     9   
289129  289129     571255    82494L  2011     10    5    17   
289134  289134     571255     82482  2011     10    5    17   
289142  289142     571255    72802B  2011     10    5    17   
289143  289143     571255    72802C  2011     10    5    17   
295956  295956     571851     23167  2011     10    3    13   
312655  312655     573362     22598  2011     10    7    13   
312656  312656     573362     22597  2011     10    7    13   
312657  312657     573362     22599  2011     10    7    13   
366326  366326     578321     20728  2011     11    3    16   
366327  366327     578321    84997B  2011     11    3    16   
366328  366328     578321    84997D  2011     11    3    16   
366329  366329     578321    84997C  2011     11    3    16   

                                Description  Quantity          InvoiceDate  \
4250                 lunch bag suki  design        70  2010-12-03 12:20:00   
4251                lunch bag pink polkadot       100  2010-12-03 12:20:00   
4252                lunch bag  black skull.        60  2010-12-03 12:20:00   
4253                lunch bag red retrospot        70  2010-12-03 12:20:00   
4254                    lunch bag cars blue       100  2010-12-03 12:20:00   
32454     red 3 piece retrospot cutlery set       144  2011-01-11 11:39:00   
32455     pink 3 piece polkadot cutlery set       360  2011-01-11 11:39:00   
32456     blue 3 piece polkadot cutlery set       360  2011-01-11 11:39:00   
32457    green 3 piece polkadot cutlery set       144  2011-01-11 11:39:00   
91155                 lunch bag suki design       150  2011-03-28 16:41:00   
91156               lunch bag pink polkadot       150  2011-03-28 16:41:00   
91159                   lunch bag cars blue       150  2011-03-28 16:41:00   
174442              lunch bag pink polkadot       100  2011-07-04 13:19:00   
174443              lunch bag  black skull.       100  2011-07-04 13:19:00   
174444              lunch bag red retrospot       100  2011-07-04 13:19:00   
174445                  lunch bag cars blue       100  2011-07-04 13:19:00   
174446      childrens cutlery retrospot red        72  2011-07-04 13:19:00   
174447      childrens cutlery polkadot pink       360  2011-07-04 13:19:00   
174448

#### Select all free orders

In [101]:
orders[orders['amount_spent']==0]

index  InvoiceNo StockCode  year  month  day  hour  \
6914      6914     537197     22841  2010     12    7    14   
22539    22539     539263     22580  2010     12    4    14   
25379    25379     539722     22423  2010     12    2    13   
29080    29080     540372     22090  2011      1    4    16   
29082    29082     540372     22553  2011      1    4    16   
34494    34494     541109     22168  2011      1    4    15   
53788    53788     543599    84535B  2011      2    4    13   
85671    85671     547417     22062  2011      3    3    10   
92875    92875     548318     22055  2011      3    3    12   
97430    97430     548871     22162  2011      4    1    14   
107192  107192     550188     22636  2011      4    4    18   
128641  128641     553000     47566  2011      5    4    15   
137127  137127     554037     22619  2011      5    5    14   
194648  194648     561284     22167  2011      7    2    12   
197173  197173     561669     22960  2011      7    4    17   
199711  199711     561916         M  2011      8    1    11   
208950  208950     562973     23157  2011      8    4    11   
223931  223931     564651     23270  2011      8    5    14   
223932  223932     564651     23268  2011      8    5    14   
223933  223933     564651     22955  2011      8    5    14   
223934  223934     564651     21786  2011      8    5    14   
256263  256263     568158      PADS  2011      9    7    12   
259399  259399     568384         M  2011      9    2     9   
273926  273926     569716     22778  2011     10    4     8   
286507  286507     571035         M  2011     10    4    12   
307092  307092     572893     21208  2011     10    3    14   
319882  319882     574138     23234  2011     11    4    11   
320049  320049     574175     22065  2011     11    4    11   
320407  320407     574252         M  2011     11    4    13   
322279  322279     574469     22385  2011     11    5    11   
328385  328385     574879     22625  2011     11    1    13   
329006  329006     574920     22899  2011     11    1    16   
329007  329007     574920     23480  2011     11    1    16   
334242  334242     575579     22437  2011     11    4    11   
334243  334243     575579     22089  2011     11    4    11   
352746  352746     577129     22464  2011     11    4    19   
353097  353097     577168         M  2011     11    5    10   
354170  354170     577314     23407  2011     11    5    13   
358629  358629     577696         M  2011     11    1    11   
370473  370473     578841     84826  2011     11    5    15   

                                Description  Quantity          InvoiceDate  \
6914           round cake tin vintage green         1  2010-12-05 14:02:00   
22539          advent calendar gingham sack         4  2010-12-16 14:36:00   
25379              regency cakestand 3 tier        10  2010-12-21 13:45:00   
29080               paper bunting retrospot        24  2011-01-06 16:41:00   
29082                plasters in tin skulls        24  2011-01-06 16:41:00   
34494          organiser wood antique white         1  2011-01-13 15:10:00   
53788          fairy cakes notebook a6 size        16  2011-02-10 13:08:00   
85671   ceramic bowl with love heart design        36  2011-03-23 10:25:00   
92875    mini cake stand  hanging strawbery         5  2011-03-30 12:45:00   
97430           heart garland rustic padded         2  2011-04-04 14:42:00   
107192   childs breakfast set circus parade         1  2011-04-14 18:57:00   
128641                        party bunting         4  2011-05-12 15:21:00   
137127            set of 6 soldier skittles        80  2011-05-20 14:13:00   
194648            oval wall mirror diamante         1  2011-07-26 12:24:00   
197173             jam making set with jars        11  2011-07-28 17:09:00   
199711                               manual         1  2011-08-01 11:44:00   
208950            set of 6 nativity magnets       240  2011-08-11 11:42:00   
223931      set of 2 

#### Select all orders that are 'lunch bag'
#### Hint: Use string functions

In [108]:
orders[orders['Description'].str.contains('lunch bag')]

index  InvoiceNo StockCode  year  month  day  hour  \
93          93     536378     20725  2010     12    3     9   
172        172     536385     22662  2010     12    3     9   
354        354     536401     22662  2010     12    3    11   
359        359     536401     20725  2010     12    3    11   
360        360     536401     22382  2010     12    3    11   
361        361     536401     20726  2010     12    3    11   
362        362     536401     22384  2010     12    3    11   
402        402     536404     22383  2010     12    3    11   
403        403     536404     20728  2010     12    3    11   
404        404     536404     20727  2010     12    3    11   
532        532     536412     20728  2010     12    3    11   
533        533     536412     22382  2010     12    3    11   
535        535     536412     22662  2010     12    3    11   
536        536     536412     22383  2010     12    3    11   
537        537     536412     20727  2010     12    3    11   
605        605     536412     20725  2010     12    3    11   
1414      1414     536542     22382  2010     12    3    14   
1415      1415     536542     20727  2010     12    3    14   
1416      1416     536542     22383  2010     12    3    14   
1748      1748     536576     20727  2010     12    3    16   
1758      1758     536577     20725  2010     12    3    16   
1771      1771     536579     20727  2010     12    3    16   
1926      1926     536597     22384  2010     12    3    17   
1935      1935     536597     20726  2010     12    3    17   
2334      2334     536634     20727  2010     12    4    11   
2529      2529     536643     22662  2010     12    4    11   
2676      2676     536698     22382  2010     12    4    12   
3059      3059     536754     22384  2010     12    4    14   
3060      3060     536754     20727  2010     12    4    14   
3064      3064     536754     22383  2010     12    4    14   
...        ...        ...       ...   ...    ...  ...   ...   
396300  396300     581412     20726  2011     12    4    14   
396301  396301     581412     20725  2011     12    4    14   
396302  396302     581412     23583  2011     12    4    14   
396303  396303     581412     20727  2011     12    4    14   
396307  396307     581412     23583  2011     12    4    14   
396308  396308     581412     20727  2011     12    4    14   
396408  396408     581413     23208  2011     12    4    14   
396414  396414     581413     22662  2011     12    4    14   
396416  396416     581413     22382  2011     12    4    14   
396419  396419     581413     20727  2011     12    4    14   
396530  396530     581418     23681  2011     12    4    15   
396572  396572     581420     23207  2011     12    4    15   
396573  396573     581420     20726  2011     12    4    15   
396574  396574     581420     23209  2011     12    4    15   
396575  396575     581420     23208  2011     12    4    15   
396576  396576     581420     22382  2011     12    4    15   
396721  396721     581426     23206  2011     12    4    15   
396970  396970     581450     20725  2011     12    4    17   
396971  396971     581450     23206  2011     12    4    17   
396972  396972     581450     23209  2011     12    4    17   
397006  397006     581451     23583  2011     12    4    17   
397038  397038     581453     22383  2011     12    4    18   
397047  397047     581453     23583  2011     12    4    18   
397255  397255     581473     23207  2011     12    4    19   
397298  397298     581474     20726  2011     12    4    20   
397465  397465     581486     23207  2011     12    5     9   
397713  397713     581538     20727  2011     12    5    11   
397714  397714     581538     20725  2011     12    5    11   
397877  397877     581581     23681  2011     12    5    12   
397880  397880     581583     20725  2011     12    5    12   

                          Description  Quantity          InvoiceDate  \
93            lunch bag red re

#### Select all orders that are made in 2011 and are 'lunch bag'  

In [109]:
orders[(orders['Description'].str.contains('lunch bag')) & (orders['year']==2011)]

index  InvoiceNo StockCode  year  month  day  hour  \
26340    26340     540015     20725  2011      1    2    11   
26341    26341     540015     20726  2011      1    2    11   
26512    26512     540023     22382  2011      1    2    12   
26513    26513     540023     20726  2011      1    2    12   
26860    26860     540098     22384  2011      1    2    15   
26861    26861     540098     22383  2011      1    2    15   
26862    26862     540098     22382  2011      1    2    15   
26863    26863     540098     20727  2011      1    2    15   
26864    26864     540098     20727  2011      1    2    15   
26902    26902     540099     20725  2011      1    2    16   
26903    26903     540099     20726  2011      1    2    16   
27118    27118     540157     22384  2011      1    3    11   
27242    27242     540168     22383  2011      1    3    12   
27360    27360     540176     22382  2011      1    3    12   
27361    27361     540176     20725  2011      1    3    12   
27727    27727     540247     22383  2011      1    3    15   
27758    27758     540247     20728  2011      1    3    15   
27760    27760     540247     22382  2011      1    3    15   
28100    28100     540268     20725  2011      1    4    11   
28136    28136     540268     20726  2011      1    4    11   
28137    28137     540268     20728  2011      1    4    11   
28138    28138     540268     22384  2011      1    4    11   
28183    28183     540275     20728  2011      1    4    12   
28188    28188     540275     20726  2011      1    4    12   
28189    28189     540275     20727  2011      1    4    12   
28201    28201     540275     20725  2011      1    4    12   
28218    28218     540276     20728  2011      1    4    12   
28219    28219     540276     20727  2011      1    4    12   
28220    28220     540276     20726  2011      1    4    12   
28221    28221     540276     20725  2011      1    4    12   
...        ...        ...       ...   ...    ...  ...   ...   
396300  396300     581412     20726  2011     12    4    14   
396301  396301     581412     20725  2011     12    4    14   
396302  396302     581412     23583  2011     12    4    14   
396303  396303     581412     20727  2011     12    4    14   
396307  396307     581412     23583  2011     12    4    14   
396308  396308     581412     20727  2011     12    4    14   
396408  396408     581413     23208  2011     12    4    14   
396414  396414     581413     22662  2011     12    4    14   
396416  396416     581413     22382  2011     12    4    14   
396419  396419     581413     20727  2011     12    4    14   
396530  396530     581418     23681  2011     12    4    15   
396572  396572     581420     23207  2011     12    4    15   
396573  396573     581420     20726  2011     12    4    15   
396574  396574     581420     23209  2011     12    4    15   
396575  396575     581420     23208  2011     12    4    15   
396576  396576     581420     22382  2011     12    4    15   
396721  396721     581426     23206  2011     12    4    15   
396970  396970     581450     20725  2011     12    4    17   
396971  396971     581450     23206  2011     12    4    17   
396972  396972     581450     23209  2011     12    4    17   
397006  397006     581451     23583  2011     12    4    17   
397038  397038     581453     22383  2011     12    4    18   
397047  397047     581453     23583  2011     12    4    18   
397255  397255     581473     23207  2011     12    4    19   
397298  397298     581474     20726  2011     12    4    20   
397465  397465     581486     23207  2011     12    5     9   
397713  397713     581538     20727  2011     12    5    11   
397714  397714     581538     20725  2011     12    5    11   
397877  397877     581581     23681  2011     12    5    12   
397880  397880     581583     20725  2011     12    5    12   

                          Description  Quantity          InvoiceDate  \
26340         lunch bag red re

#### Show the frequency distribution of the amount spent in Spain.

In [110]:
orders['amount_spent'].value_counts()

15.00      20082
19.80      11033
17.70       9174
16.50       8490
10.20       8028
1.25        7552
3.75        6847
4.95        6507
1.65        6228
10.50       5550
20.80       5524
2.95        5432
5.04        5284
2.50        5242
30.00       5226
7.50        5163
3.30        5063
8.50        4953
17.00       4899
9.90        4810
17.40       4394
14.85       4246
2.10        4041
15.30       3951
5.90        3669
2.55        3661
15.90       3562
13.20       3517
10.08       3248
0.85        3143
           ...  
175.60         1
282.24         1
18.48          1
2118.74        1
449.28         1
1187.20        1
1910.40        1
125.55         1
716.40         1
1505.28        1
6.48           1
25.08          1
323.40         1
1024.00        1
72.93          1
892.50         1
425.76         1
362.70         1
480.24         1
13.34          1
101.60         1
725.04         1
120.45         1
159.50         1
436.80         1
1275.00        1
18.56          1
1.32          

#### Select all orders made in the month of August

In [112]:
orders[orders['month']==8]

index  InvoiceNo StockCode  year  month  day  hour  \
199475  199475     561904     22075  2011      8    1     8   
199476  199476     561904    85049E  2011      8    1     8   
199477  199477     561905     21385  2011      8    1     9   
199478  199478     561905    84970L  2011      8    1     9   
199479  199479     561905    84970S  2011      8    1     9   
199480  199480     561905     84978  2011      8    1     9   
199481  199481     561905     21754  2011      8    1     9   
199482  199482     561905     23226  2011      8    1     9   
199483  199483     561905     21715  2011      8    1     9   
199484  199484     561905     21716  2011      8    1     9   
199485  199485     561905     21746  2011      8    1     9   
199486  199486     561905     21755  2011      8    1     9   
199487  199487     561906     22367  2011      8    1    10   
199488  199488     561906     20914  2011      8    1    10   
199489  199489     561906     21240  2011      8    1    10   
199490  199490     561906     20725  2011      8    1    10   
199491  199491     561906     20726  2011      8    1    10   
199492  199492     561906     22382  2011      8    1    10   
199493  199493     561906     22629  2011      8    1    10   
199494  199494     561906     22384  2011      8    1    10   
199495  199495     561906     23206  2011      8    1    10   
199496  199496     561906     23209  2011      8    1    10   
199497  199497     561906     22898  2011      8    1    10   
199498  199498     561906     21035  2011      8    1    10   
199499  199499     561906     21217  2011      8    1    10   
199500  199500     561906     21155  2011      8    1    10   
199501  199501     561906     21154  2011      8    1    10   
199502  199502     561906     21094  2011      8    1    10   
199503  199503     561906     21080  2011      8    1    10   
199504  199504     561906     84991  2011      8    1    10   
...        ...        ...       ...   ...    ...  ...   ...   
226458  226458     565067    85183B  2011      8    3    17   
226459  226459     565067     22065  2011      8    3    17   
226460  226460     565067     22900  2011      8    3    17   
226461  226461     565067     21508  2011      8    3    17   
226462  226462     565067     22713  2011      8    3    17   
226463  226463     565067     22716  2011      8    3    17   
226464  226464     565067     21558  2011      8    3    17   
226465  226465     565067     22352  2011      8    3    17   
226466  226466     565067    84569A  2011      8    3    17   
226467  226467     565067    84569D  2011      8    3    17   
226468  226468     565067    84559A  2011      8    3    17   
226469  226469     565067    84559B  2011      8    3    17   
226470  226470     565067    84558A  2011      8    3    17   
226471  226471     565067     22189  2011      8    3    17   
226472  226472     565067     22188  2011      8    3    17   
226473  226473     565067     22224  2011      8    3    17   
226474  226474     565067     21248  2011      8    3    17   
226475  226475     565067     22109  2011      8    3    17   
226476  226476     565067     22107  2011      8    3    17   
226477  226477     565067    84519B  2011      8    3    17   
226478  226478     565067     21559  2011      8    3    17   
226479  226479     565067     22352  2011      8    3    17   
226480  226480     565067     21558  2011      8    3    17   
226481  226481     565067     22616  2011      8    3    17   
226482  226482     565067     16048  2011      8    3    17   
226483  226483     565067     22644  2011      8    3    17   
226484  226484     565067     22645  2011      8    3    17   
226485  226485     565067     22637  2011      8    3    17   
226486  226486     565067     22646  2011      8    3    17   
226487  226487     565067     22741  2011      8    3    17   

                                Description  Quantity          InvoiceDate  \
199475          6 ribbon

#### Select how many orders are made by countries in the month of August
##### Hint: Use value_counts()

In [114]:
orders[orders['month']==8]['Country'].value_counts()

United Kingdom     23105
Germany              795
EIRE                 593
France               569
Netherlands          280
Switzerland          267
Spain                252
Belgium              194
Israel               171
Channel Islands      140
Australia            107
Italy                 95
Austria               88
Norway                77
Finland               61
Malta                 55
Portugal              41
Sweden                40
Unspecified           23
Iceland               22
Poland                17
Denmark               16
Canada                 5
Name: Country, dtype: int64

#### What's the  average amount of money spent by country

In [119]:
orders.groupby('Country').mean()[['amount_spent']]

amount_spent
Country                           
Australia               116.895620
Austria                  25.624824
Bahrain                  32.258824
Belgium                  20.283772
Brazil                   35.737500
Canada                   24.280662
Channel Islands          27.340160
Cyprus                   22.134169
Czech Republic           33.069600
Denmark                  49.882474
EIRE                     36.687745
European Community       21.670833
Finland                  32.913985
France                   25.056827
Germany                  25.311562
Greece                   32.831172
Iceland                  23.681319
Israel                   29.119718
Italy                    23.064960
Japan                   116.561900
Lebanon                  37.641778
Lithuania                47.458857
Malta                    24.335625
Netherlands             120.798282
Norway                   33.736418
Poland                   22.226212
Portugal                 22.872702
RSA                      17.281207
Saudi Arabia             16.213333
Singapore                95.852658
Spain                    24.779521
Sweden                   85.096075
Switzerland              30.642752
USA                      20.002179
United Arab Emirates     27.974706
United Kingdom           20.625073
Unspecified              10.930615

#### What's the most expensive item?

In [120]:
orders[orders['UnitPrice']==orders['UnitPrice'].max()]

index  InvoiceNo StockCode  year  month  day  hour Description  \
118352  118352     551697      POST  2011      5    2    13     postage   

        Quantity          InvoiceDate  UnitPrice  CustomerID         Country  \
118352         1  2011-05-03 13:46:00    8142.75       16029  United Kingdom   

        amount_spent  
118352       8142.75

#### What was the average amount spent per year ?

In [123]:
orders.groupby('year').mean()[['amount_spent']]

amount_spent
year              
2010     21.892733
2011     22.430074